### ISSUE TITLE
Improve handling of incorrect calls to Assert.AreSame

### ISSUE THREAD

**Issue by bjornhellander** [microsoft/testfx#4352]
It is currently possible to call Assert.AreSame with value ~reference~ types. The assertion will fail when the test is run, but it would be better if
1) this was not possible at all (by constraining the generic type parameter), or
2) an analyzer could warn immediately about calls that will fail

Not a big problem since the test will fail, but would be nice since it would shorten the time before the mistake was discovered.

**Comment by Evangelink** [comment]
Hi @bjornhellander ,

Would you mind giving me an example of what is not working?
AreSame api is basically just a wrapper around `object.Referencequal` so being able to pass reference seems fine by me.

**Comment by bjornhellander** [comment]
Yeah, I messed up. I have edited the description now. I meant **value** types. So if you mistakenly call AreSame with arguments of a value type, the assert will fail and inform you that you have made a mistake, but it would be nice if the mistake could be immediately apparent instead.

**Comment by Evangelink** [comment]
Makes sense!

The goal is to add an analyzer (enabled by default + severity = warning) that would flag calls to `Assert.AreSame` where 1 side is a value type.
